src/sys/petsc_interface.cxx

```diff
@@ -0,0 +1,248 @@
+      if (it.ind == localmesh->xstart) indices2D(it.ind - 1, localmesh->ystart - 1) = counter++;
+      if (it.ind == localmesh->xend) indices2D(it.ind + 1, localmesh->ystart - 1) = counter++;
+    indices2D(it.ind, localmesh->ystart - 1) = counter++;
+  }
+  for (RangeIterator it=localmesh->iterateBndryUpperY(); !it.isDone(); it++) {
+      if (it.ind == localmesh->xstart) indices2D(it.ind - 1, localmesh->yend + 1) = counter++;
+      if (it.ind == localmesh->xend) indices2D(it.ind + 1, localmesh->ystart + 1) = counter++;
+    indices2D(it.ind, localmesh->yend + 1) = counter++;
+  }
+  BOUT_FOR(i, localmesh->getRegion2D("RGN_NOY")) {
+    if ((i.x() >= localmesh->xstart && i.x() <= localmesh->xend) ||
+	(i.x() < localmesh->xstart && localmesh->firstX()) ||
+	(i.x() > localmesh->xend && localmesh->lastX())) {
+      indices2D[i] = counter++;
+    }
+  }
+
+  // Set up the Perp indices; will these work in general or will
+  // different ones be needed for each value of y?
+  counter = localmesh->globalStartIndexPerp();
+  BOUT_FOR(i, localmesh->getRegionPerp("RGN_NOY")) {
+    if ((i.x() >= localmesh->xstart && i.x() <= localmesh->xend) ||
+	(i.x() < localmesh->xstart && localmesh->firstX()) ||
+	(i.x() > localmesh->xend && localmesh->lastX())) {
+      indicesPerp[i] = counter++;
+    }
+  }
+}
+
+// PetscVectorElement implementation
+
+PetscVectorElement::PetscVectorElement(Vec* vector, int index) :
+  petscVector(vector), petscIndex(index) {
+}
+
+BoutReal PetscVectorElement::operator=(BoutReal val) {
+  auto status = VecSetValues(*petscVector, 1, &petscIndex, &val, INSERT_VALUES);
+  delete this;
+  if (status != 0) {
+    throw BoutException("Error when setting elements of a PETSc vector.");
+  }
+  return val;
+}
+
+BoutReal PetscVectorElement::operator+=(BoutReal val) {
+  auto status = VecSetValues(*petscVector, 1, &petscIndex, &val, ADD_VALUES);
+  delete this;
+  if (status != 0) {
+    throw BoutException("Error when setting elements of a PETSc vector.");
+  }
+  return val;
+}
+
+PetscVectorElement& PetscVectorElement::newElement(Vec* vector, PetscInt index) {
+  PetscVectorElement* ve = new PetscVectorElement(vector, index);
+  return *ve;
+}
+
+
+// PetscMatrixElement implementation
+
+PetscMatrixElement::PetscMatrixElement(Mat* matrix, PetscInt row,
+    std::vector<PetscInt> p, std::vector<BoutReal> w) : petscMatrix(matrix),
+    petscRow(row),  positions(p), weights(w) {
+}
+
+BoutReal PetscMatrixElement::operator=(BoutReal val) {
+  setValues(val, INSERT_VALUES);
+  delete this;
+  return val;
+}
+
+BoutReal PetscMatrixElement::operator+=(BoutReal val) {
+  setValues(val, ADD_VALUES);
+  delete this;
+  return val;
+}
+
+void PetscMatrixElement::setValues(BoutReal val, InsertMode mode) {
+  int num = positions.size();
+  ASSERT3(num > 0);
+  PetscInt* columns = new PetscInt[num];
+  PetscScalar* values = new PetscScalar[num];
+  for (int i = 0; i < num; i++) {
+    columns[i] = positions[i];
+    values[i] = weights[i]*val;
+  }
```

Comment:
Could just use `std::vector` and `std::transform` here.

Reply:
Hadn't thought to do that, as PETSc is expecting raw pointers, but see it is possible using std::vector::data() method. I've now switched.